rfcs/53-generalized-hierarchy.md

```diff
@@ -0,0 +1,273 @@
+a single component that creates a top-down forest of entity trees, where each
+entity forms a doubly-linked list of siblings. It may also cache infromation
+about the component's position within the hierarchy like it's depth.
+
+Mutations to the hierarchy: de-parenting, re-parenting, re-ordering of children,
+etc. can only be done via Commands. This enforces a globally consistent view at
+any given time. Delaying any modifications to a global synchronization point,
+much like component removal. The `Hierarchy` component is not publicly
+constructible and must be made via the dedicated hierarchy management commnds.
+
+To make querying for specific parts of the hierarchy easier, ZST marker
+components will be added and removed to make querying for specific cases easier,
+and will be kept in sync with the hierarchy state: `Parent`, `Child`,
+`FirstSibling`, `LastSibling`. These can be used in conjunction with ECS Query
+filters to speed up finding targetted use cases. For example, `Without<Child>`
+can be used to query for roots, or `With<Parent>` can be used to find all of the
+non-leaf entities.
+
+To signal changes in the hierachy, a `HierarchyEvent` event type is registered
+and any changes made to the hierarchy is published as an event.
+
+## Implementation strategy
+The core of this proposal is the `Hierarchy` component, which roughly looks like
+this:
+
+```rust
+#[derive(Component)]
+pub struct Hierarchy {
+  depth: u16,
+  parent: Option<Entity>,
+  prev: Option<Entity>,
+  next: Option<Entity>
+  first_child: Option<Entity>,
+}
+```
+
+This type has no public constructors and has no public APIs that allow for
+mutation of any of the internal data, only getters. The consistency between
+multiple instances of the component is maintained by a set of dedicated hierachy
+management commands. The following should be added as extensions to
+EntityCommands:
+
+  - `set_parent(Option<Entity>)` - sets the parent of a given entity.
+      - If `Hierarchy` is on neither parent or child, it will be added and both
+        will be atomically updated to reflect this change.
+      - If the parent already had children, the new child will be prepended at
+        the start of the linked list, and the sibling pointers of both will be
+        updated.
+      - If the child already had a parent, that previous parent will be updated
+        too.
+      - A HierarchyEvent is added to the queue to reflect this update.
+  - `move_to_first(Entity)` - Moves one of the children to the front.
+  - `move_to_last(Entity)` - Moves one of the children to the end.
+  - `remove_all_children()` - Disconnects all children from a parent. All
+    children become roots.
+      - A HierarchyEvent is added to the queue to reflect this update.
+
+All of these commands will also update the four marker components appropriately.
+Optionally, we could also mark these components as changed (even though they're
```

Comment:
Hmm. I think I like this.

In general, stuff like this belongs in the "Open Questions" section.

Reply:
Upon reflection: I think we should kill these marker components entirely.

Reply:
Removed the mention of these given the recent optimization insights.